pkg/apis/autopilot/v1alpha1/actionapproval.go

```diff
@@ -0,0 +1,72 @@
+package v1alpha1
+
+import (
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	// ActionApprovalResourcePlural is the name of the plural ActionApproval objects
+	ActionApprovalResourcePlural = "actionapprovals"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ActionApproval ia used for approval of actions to be taken by Autopilot as part of a rule
+type ActionApproval struct {
+	meta.TypeMeta   `json:",inline"`
+	meta.ObjectMeta `json:"metadata,omitempty"`
+	Spec            ActionApprovalSpec   `json:"spec"`
+	Status          ActionApprovalStatus `json:"status"`
+}
+
+// ActionApprovalSpec is the spec for an autopilot action approval
+type ActionApprovalSpec struct {
+	// ApprovalState is the current approval state of the action
+	ApprovalState ActionApprovalState `json:"approvalState,omitempty"`
+	// Actions are the actions that needs/needed approval
+	Actions []*RuleAction `json:"actions,omitempty"`
+}
+
+// ActionApprovalStatus is current status of an autopilot action approval
+type ActionApprovalStatus struct {
+	// Rule is the parent autopilot rule that resulted in this action approval
+	Rule types.NamespacedName
+	// LastProcessTimestamp was the last time this approval was processed
+	LastProcessTimestamp meta.Time `json:"lastProcessTimestamp"`
+	// ActionPreviews provides a dry-run preview of the side-effects of the actions
+	ActionPreviews []*AutopilotActionPreview `json:"actionPreviews,omitempty"`
+}
+
+// AutopilotActionPreview is a preview of an action and the expected result of it before it gets executed
+type AutopilotActionPreview struct {
+	// Action is the action spec that's going to get executed
+	Action *RuleAction `json:"action,omitempty"`
+	// ExpectedResult is a user friendly description of the outcome of executing the action
+	ExpectedResult string `json:"expectedResult,omitempty"`
```

Comment:
Should we make this ExpectedResult a struct and then add a Message as a string. In this way it is more extensible if we need to have well defined Results.

Reply:
Good idea. Done. Also added a String() function which currently just returns the Message. If we add fields in future, that can pretty format the result.